### PR TITLE
 Enhance NuGet.Protocol.Catalog to support other client implementations

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/6444
Progress on https://github.com/NuGet/NuGetGallery/issues/6445

`ISimpleHttpClient` will be used by a future `RegistrationClient` to read registration blobs.

Also, add `IsSemVer2` extension method to package details catalog leaves.

### CODE REVIEW

File: src/NuGet.Protocol.Catalog/SimpleHttpClient.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace NuGet.Protocol.Catalog
+{
+    public class SimpleHttpClient : ISimpleHttpClient
+    {
+        private static readonly JsonSerializer JsonSerializer = CatalogJsonSerialization.Serializer;
```
Comment (loic-sharma) [inline src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: >JsonSerializer [](start = 47, length = 14)

If it wasn't for this `JsonSerializer`, `SimpleHttpClient` wouldn't be catalog specific. Instead of having a static property, could we instead accept a `JsonSerializer` in the constructor? What are the tradeoffs?

File: src/NuGet.Protocol.Catalog/SimpleHttpClient.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public SimpleHttpClient(HttpClient httpClient, ILogger<SimpleHttpClient> logger)
+            : this(httpClient, (ILogger)logger)
+        {
+        }
+
+        internal SimpleHttpClient(HttpClient httpClient, ILogger logger)
```
Comment (loic-sharma) [inline src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: >SimpleHttpClient [](start = 17, length = 16)

What's this constructor for?

Comment (joelverhagen) [reply src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: Catalog client's existing ctor (to avoid API breaking changes).

File: src/NuGet.Protocol.Catalog/SimpleHttpClient.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: I was favoring a simple client that should work on all V3 JSON. I think the name of the static class `CatalogJsonSerialization` is what is confusing.

Perhaps renaming it to `NuGetJsonSerialization` would clear this up?

File: src/NuGet.Protocol.Catalog/SimpleHttpClient.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace NuGet.Protocol.Catalog
+{
+    public class SimpleHttpClient : ISimpleHttpClient
```
Comment (chenriksson) [inline src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: nit: why not `JsonHttpClient` or similar to reflect the JSON deserialization that the class does?

File: src/NuGet.Protocol.Catalog/Models/ModelExtensions.cs [hunk 4]
```diff
@@ -166,5 +166,47 @@ public static bool IsListed(this PackageDetailsCatalogLeaf leaf)
...
             return leaf.Published.Year != 1900;
         }
+
+        /// <summary>
+        /// Determines if the provied package details leaf represents a SemVer 2.0.0 package. A package is considered
+        /// SemVer 2.0.0 if it's version is SemVer 2.0.0 or one of its dependency version ranges is SemVer 2.0.0.
+        /// </summary>
+        /// <param name="leaf">The catalog leaf.</param>
+        /// <returns>True if the package is SemVer 2.0.0.</returns>
+        public static bool IsSemVer2(this PackageDetailsCatalogLeaf leaf)
```
Comment (chenriksson) [inline src/NuGet.Protocol.Catalog/Models/ModelExtensions.cs]: Assume this will be used in a separate PR, for bug fix?

Comment (joelverhagen) [reply src/NuGet.Protocol.Catalog/Models/ModelExtensions.cs]: Separate PR.

File: src/NuGet.Protocol.Catalog/SimpleHttpClient.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Protocol.Catalog/SimpleHttpClient.cs]: My hope is that other convenience methods could be added here as well. For example, fetching the .nuspec as `XDocument` or a .nupkg as a local temp file. `Json` is just the first convenience needed.
